File: python/tree_heap.py
import heapq
from dataclasses import dataclass
from typing import Optional, List


@dataclass
class Node:
    value: int
    left: Optional["Node"] = None
    right: Optional["Node"] = None


class BinarySearchTree:
    def __init__(self):
        self.root: Optional[Node] = None

    def insert(self, value: int):
        self.root = self._insert(self.root, value)

    def _insert(self, node: Optional[Node], value: int) -> Node:
        if node is None:
            return Node(value)
        if value < node.value:
            node.left = self._insert(node.left, value)
        else:
            node.right = self._insert(node.right, value)
        return node

    def inorder(self) -> List[int]:
        res: List[int] = []

        def dfs(node: Optional[Node]):
            if not node:
                return
            dfs(node.left)
            res.append(node.value)
            dfs(node.right)

        dfs(self.root)
        return res


class MinHeap:
    def __init__(self):
        self._h = []

    def push(self, v: int):
        heapq.heappush(self._h, v)

    def pop(self) -> int:
        return heapq.heappop(self._h)

    def peek(self) -> int:
        return self._h[0]

    def is_empty(self) -> bool:
        return len(self._h) == 0

File: golang/tree/tree_heap.go
package tree

import "container/heap"

type Node struct {
	Val         int
	Left, Right *Node
}

type BST struct {
	Root *Node
}

func (b *BST) Insert(v int) {
	b.Root = insertNode(b.Root, v)
}

func insertNode(n *Node, v int) *Node {
	if n == nil {
		return &Node{Val: v}
	}
	if v < n.Val {
		n.Left = insertNode(n.Left, v)
	} else {
		n.Right = insertNode(n.Right, v)
	}
	return n
}

func (b *BST) Inorder() []int {
	res := []int{}
	var dfs func(*Node)
	dfs = func(n *Node) {
		if n == nil {
			return
		}
		dfs(n.Left)
		res = append(res, n.Val)
		dfs(n.Right)
	}
	dfs(b.Root)
	return res
}

type IntMinHeap []int

func (h IntMinHeap) Len() int           { return len(h) }
func (h IntMinHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h IntMinHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *IntMinHeap) Push(x any)        { *h = append(*h, x.(int)) }
func (h *IntMinHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

func HeapPush(h *IntMinHeap, v int) {
	heap.Push(h, v)
}

func HeapPop(h *IntMinHeap) int {
	return heap.Pop(h).(int)
}
